lib/lexer/lexer.cpp

```diff
@@ -0,0 +1,484 @@
+		} else {
+			delimit(Token{Token::Type::Token, WordToken{std::move(state_data.word)}});
+		}
+		state_data.word.clear();
+	}
+	return state;
+}
+
+Lexer::State Lexer::operatorState() {
+	char ch = chars.peek();
+
+	if (isBackslash(ch)) {
+		state_data.pushState(State::Operator);
+		return State::Backslash;
+	}
+
+	int matches;
+	if (!isOperatorCharacter(ch)) {
+		matches = 0;
+	} else {
+		matches = Token::prefixOperatorMatches(state_data.word + ch);
+	}
+
+	if (matches >= 1) {
+		// The characters we have now are part of 1 or multiple operators
+		// don't delimit yet, because this does not tell us that it matches an entire operator
+		state_data.word.push_back(ch);
+		chars.consume();
+		return State::Operator;
+	}
+	delimitOperator();
+	return State::Empty;
+}
+
+// Previously it did match a prefix of a valid operator
+// but with the additional character it doesn't;
+// - find out what the previous operator was
+void Lexer::delimitOperator() {
+	auto result = Token::exactOperatorType(state_data.word);
+	D_ASSERT(result.has_value());
+	delimit(Token {result.value(), OperatorToken()});
+	state_data.word.clear();
+}
+
+Lexer::State Lexer::commentState() {
+	char ch = chars.peek();
+	if (isNewline(ch) || ch == EOF) {
+		return State::Empty;
+	}
+	chars.consume();
+	return State::Comment;
+}
+
+// Precondition: backslash is NOT consumed. Previous state is either WORD or OPERATOR
+Lexer::State Lexer::backslashState() {
+	char ch = chars.consume();
+	D_ASSERT(isBackslash(ch));
+	ch = chars.peek();
+	if (ch == EOF) {
+		// todo: determine behavior
+		// throw std::runtime_error("encountered EOF in backslash lexing");
+		return state_data.popState();
+	}
+	if (isNewline(ch)) {
+		// Special case: `\\n`: remove backslash and newline from the input stream and continue where we left off.
+		chars.unconsume();
+		// Todo: if the NEXT character IS another newline, then the backslash _IS_ added to the history. Maybe just a
+		// bash/zsh behavior?
+		chars.remove();
+		chars.remove();
+		return state_data.popState();
+	}
+
+	// This is required to be here in the case of something like `>\\n>`, which results in the next token being used as
+	// an operator. Since we have not encountered a newline, that scenario is no longer possible and the literal value
+	// is used as part of a word, so we delimit the operator at this point.
+	if (state_data.previousState() == State::Operator) {
+		delimitOperator();
+	}
+
+	state_data.word.push_back('\\');
+	state_data.word.push_back(ch);
+	chars.consume();
+	return state_data.popState();
+}
+
+/*
+Preconditions:
+	- Starting single quote is consumed
+	- Previous state is on the stack
+No transitions, everything is literal.
+*/
+Lexer::State Lexer::singleQuoteState() {
+	while (true) {
+		char ch = chars.consume();
+		if (ch == EOF) {
+			// todo: determine behavior
+			// throw std::runtime_error("encountered EOF in single quote parsing");
+			return state_data.popState();
+		}
+		state_data.word.push_back(ch);
```

Comment:
I think this still needs backslash behavior, no?
```
➜  tmp git:(redirection) ✗ echo '\\\d'
\\d
```
Or is this zsh-only behavior?

Reply:
Single quotes should preserve all literals so not sure what's happening there
https://pubs.opengroup.org/onlinepubs/009695399/utilities/xcu_chap02.html#tag_02_02_02

bash:
```bash
$ echo '\\\d'
\\\d
```

Reply:
If there was backslash expansion inside of single quotes then the result would also be `\d`, not `\\d`, so I'm a bit confused about the zsh result tbh.

Reply:
I think what it does is escape the `\`, so `\\` becomes `\`, and then the `\d` is not special so the backslash doesn't get removed

But in any case, seems to be zsh behavior

Reply:
Yeah, you're right. This will be the behavior when expanding double quotes actually